packages/react/src/emotion-element.js

```diff
@@ -49,7 +54,32 @@ export const createEmotionProps = (type: React.ElementType, props: Object) => {
   return newProps
 }
 
-const Noop = () => null
+const Insertion = ({ cache, serialized, isStringTag }) => {
+  registerStyles(cache, serialized, isStringTag)
+
+  const rules = useInsertionEffectMaybe(() =>
```

Comment:
One thing to consider - it would be cool to flatten the React tree cause this `Insertion` component is not needed. Or rather, it is **somewhat** needed:
- it creates a "slot" so we avoid SSR mismatches when `useId` is used. But that is only a requirement when relying on 0-config and I think that a lot of people are not actually using this feature. With most SSR-oriented APIs we don't have to render `style` elements "inline" and in those setups there is no need for this extra "slot"
- it preserves injection order. We know that injection order is not something that people should rely on but we'd like to make it as easy as possible for them to migrate

It seems that we can't easily work around the second "requirement' unless we provide a way for people to configure `injectWhenever`. I think the plan is to just make this a default in the next major version though - so the question is, should we allow people to flatten their React trees with a flag or something?

Reply:
I think for now, let's leave it out.